zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -103,15 +127,134 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+    // - h_sig remaining bits [2*size_value + 1, 2*size_value + (digest_length-field_capacity)[
+    // - nullifierS remaining bits [2*size_value + (digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs)*(digest_length-field_capacity)[
+    // - commitmentS remaining bits [2*size_value + (1+NumInputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity)[
+    // - message authentication tagS remaining bits [2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+2*NumInputs+NumOutputs)*(digest_length-field_capacity)]
     // ============================================================================================ //
 
+    // This function is used to extract the public values (vpub_in and vpub_out) from the residual field element(S)
+    function assemble_public_values(uint[] memory primary_inputs) public view returns (uint64 vpub_in, uint64 vpub_out){
+        // We know vpub_in corresponds to the first 64 bits of the first residual field element after padding.
+        // We start by computing the padding .
+        uint padding = packing_residue_length;
+        // Update the padding if the public values and remainder bits fit in one field element
+        if (2*size_value+padding*(1+2*jsIn+jsOut) < field_capacity) {
+            padding = digest_length - (2*size_value + packing_residue_length*(1+2*jsIn+jsOut));
+        }
+
+        // We retrieve the public value in, remove any extra bits (due to the padding) and inverse the bit order
+        bytes32 vpub_bytes = (bytes32(primary_inputs[1 + 1 + 2*jsIn + jsOut]) << padding) >> (digest_length-size_value);
+        vpub_bytes = Bytes.swap_bit_order(vpub_bytes) >> (digest_length-size_value);
+        vpub_in = Bytes.get_int64_from_bytes8(Bytes.int256ToBytes8(uint(vpub_bytes)));
+
+        // We retrieve the public value out, remove any extra bits (due to the padding) and inverse the bit order
+        vpub_bytes = (bytes32(primary_inputs[1 + 1 + 2*jsIn + jsOut]) << (padding+size_value)) >> (digest_length-size_value);
+        vpub_bytes = Bytes.swap_bit_order(vpub_bytes) >> (digest_length-size_value);
+        vpub_out = Bytes.get_int64_from_bytes8(Bytes.int256ToBytes8(uint(vpub_bytes)));
+    }
+
+    // This function is used to reassemble hsig given the the primary_inputs
+    // To do so, we extract the remaining bits of hsig from the residual field element(S) and combine them with the hsig field element
+    function assemble_hsig(uint[] memory primary_inputs) public view returns (bytes32 hsig){
+
+        // We know hsig residual bits correspond to the 128th to 130st bits of the first residual field element after padding.
+        // We start by computing the padding.
+        uint padding = packing_residue_length;
+        // Update the padding if the public values and remainder bits fit in one field element
+        if (2*size_value+padding*(1+2*jsIn+jsOut) < field_capacity) {
```

Comment:
If thsi condition does not hold, do you not need to also adjust the index `1 + 1 + 2*jsIn + jsOut` into `primary_inputs` that the residuals get taken from?  (Similarly in other `assemble_*` functions).  Or, perhaps it's safer to just `require` that the residuals + public values don't overflow a single field element.

Reply:
Yes, I assumed that the condition holds.
(It was quite expensive to consider the case where we have more than one field element as residual bits since we have to handle 2 bytes at a time in the worst case scenario).

I am not sure where to add the following require however:
```
    require(
        2*public_value_length+packing_residue_length*(1+2*jsIn+jsOut) < field_capacity,
        "too many input and output notes considered"
    );
```

Reply:
After talking, we decided that the constructor was the best place to write this require.